msx.cpp: Made the general purpose ports (joystick/mouse) a slot device. Add Bee Card support.

wilbertpol opened the pull request.

startaq commented:
There is now `bus/msx_cart`, `bus/msx_slot` and `bus/msx_ctrl`. Maybe a subdirectory `msx` would be cleaner? I. e. `bus/msx/cart` etc.

wilbertpol commented:
> There is now `bus/msx_cart`, `bus/msx_slot` and `bus/msx_ctrl`. Maybe a subdirectory `msx` would be cleaner? I. e. `bus/msx/cart` etc.

Yes, that would be cleaner

ajrhacker commented:
I'm not very fond of putting bus slots in subdirectories, given how several connector types ended up being adopted by otherwise unrelated machines by different manufacturers. The MSX joystick ports, in particular, are practically identical to those of several other (mostly Japanese) computers, and are also largely similar to the controller ports of the Atari VCS and Sega's 8/16-bit consoles.

angelosa commented:
Ideally the connectors are the same, it's the pinout meaning that differs. If I have to trust https://github.com/mamedev/mame/blob/852e9a3703b0d134be18d0156d2385a48da3f9bc/src/mame/sharp/x68k.cpp#L508
then x68000/msx/fm towns/megadrive all runs over the same model, iirc pc8801 is the same but I don't have pinout handy atm to confirm.

wilbertpol commented:
I changed the PR a bit to add a generic DE-9 connector in bus/de9 and moved the msx controller stuff to bus/de9/msx_ctrl.
The DE-9 slot definition will be too limited for a straight drop-in of the vcs controllers though; that will have to be addressed if/when the vcs_ctrl bus gets moved.

0kmg commented on src/mame/msx/msx.cpp:
```diff
@@ -432,53 +428,16 @@ void msx_state::driver_start()
-		data |= game_port_r<1>();
-	}
+	if (BIT(m_psg_b, 6))
+		data |= bitswap<8>(m_gen_port2->read(), 4, 7, 6, 5, 3, 2, 1, 0) & 0x3f;
```
This can be simplified to `bitswap<6>(m_gen_port2->read(), 6, 5, 3, 2, 1, 0);`.

wilbertpol replied:
Thanks, fixed both of them.

0kmg commented on src/mame/msx/msx.cpp:
```diff
@@ -432,53 +428,16 @@ void msx_state::driver_start()
-		// game port 1
-		data |= game_port_r<0>();
-	}
+		data |= bitswap<8>(m_gen_port1->read(), 4, 7, 6, 5, 3, 2, 1, 0) & 0x3f;
```
Same.

cuavas commented:
I don’t think trying to merge electrically incompatible peripheral connectors is a good idea.
* The Atari and MSX controller connectors are sufficiently different in an electrically incompatible way to warrant keeping them separate.  Both common ground and +5V power are on different pins.
* The Mega Drive controller port is logically similar to an MSX controller port (six logic inputs, one logic output, +5V and common ground), but the common ground and output are on different pins.  Sega uses pin 8 for ground while MSX uses pin 9.
* The master system is a bit of an oddity, as two of the pins can be configured as inputs or outputs – it can support up to seven inputs if no outputs are needed.  The pin direction can be controlled in software.
* Commodore extended the Atari joystick interface, while other Japanese home computers followed the MSX.

I don’t think calling it DE9 is a good idea at all, since the connector was used for plenty of other *completely* incompatible things, e.g. Mac mouse port, PC RS-232 port, various RGB video standards (MDA, Hercules, CGA, EGA, and even a variant of VGA without separate colour channel grounds), CAN bus, and even Token Ring on some bridges and routers.

I think it’s probably best to call the MSX controller port `msx_ctrl` or some variant as MSX was posited as a kind of home computer standard, and the other adopters (e.g. Sharp with the X68000, and Fujitsu with the FM Towns) were following the standard.

wilbertpol commented:
Ok, I have backed out the DE9 changes. The msx controller code is now in bus/msx/ctrl again.

angelosa commented on src/devices/bus/msx/ctrl/mouse.cpp:
```diff
@@ -0,0 +1,93 @@
+			s16 mouse_x = m_port_mouse_x->read();
+			s16 mouse_y = m_port_mouse_y->read();
+
+			m_data = (u8(m_mouse_x - mouse_x) << 8) | u8(m_mouse_y - mouse_y);
```
fwiw this looks similar to what I have in [pc8801.cpp](https://github.com/mamedev/mame/blob/ce88c950567033a9f23da716a14e0a9d0862e815/src/mame/nec/pc8801.cpp#L680)

I'm pretty convinced the whole implementation should stay in a standalone `msx_ctrl` folder.

cuavas replied:
Does putting it in `bus/msx_ctrl` vs `bus/msx/ctrl` really make a meaningful difference?  Any driver (or other device) can use it wherever it sits in the source tree.

angelosa replied:
Yeah, may as well work either way really. 
Just mentioning because I eventually have to merge this in my pc88va branch.

wilbertpol replied:
> fwiw this looks similar to what I have in [pc8801.cpp](https://github.com/mamedev/mame/blob/ce88c950567033a9f23da716a14e0a9d0862e815/src/mame/nec/pc8801.cpp#L680)

That mouse implementation indeed looks the same.

wilbertpol commented:
Redoing this to workaround the conflict
